Enhance Route Message With Background and Fuel Costs

ron-murhammer opened the pull request:
Addresses point 3 of https://forums.triplea-game.org/topic/558/fuel-enhancements

**Functional Changes**
- Route movement message now has a background and displays fuel costs
- Adjusted some colors and sizes as well to improve consistency and readability

**Before**
![image](https://user-images.githubusercontent.com/1427689/36769054-a1b81330-1c07-11e8-8793-05be5044718e.png)

**After**
![image](https://user-images.githubusercontent.com/1427689/36806413-8e706466-1c85-11e8-9528-8d3b1d4e732e.png)

ssoloff commented on game-core/src/main/java/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -314,28 +323,81 @@ private void drawCurvedPath(final Graphics2D graphics, final Point2D[] points) {
+    graphics.drawRoundRect(0, 0, imageWidth - 2, rectHeight, rectHeight, rectHeight);
+
+    // Draw current movement
+    graphics.setColor(new Color(255, 120, 0));
```
Just noting that I find this foreground color a bit hard to distinguish between the background color.  It looks like it's a medium orange on sand.  Not a big deal, just wanted to mention it.

ron-murhammer replied:
I adjusted a couple of the colors so that its more of a light gray background and a darker orange.

ssoloff commented on game-core/src/main/java/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -314,28 +323,81 @@ private void drawCurvedPath(final Graphics2D graphics, final Point2D[] points) {
+      final ResourceCollection movementFuelCost, final ResourceImageFactory resourceImageFactory) {
+
+    // Create and configure image
+    final String unitMovementLeft = (maxMovement == null || maxMovement.trim().length() == 0) ? "" : "/" + maxMovement;
```
I'm not sure if `" / "` is a better separator than `"/"` here.  The latter seems a bit too cramped to me, but some may think the former is too spread out. :man_shrugging:  Regardless, it's better than the old separator:

```
"    /"
```

RoiEXLab replied:
Not that important, but maxMovement should really be an int with a value of -1 or something for the null replacement if necessary

ron-murhammer replied:
I tried that as well but felt without spaces was more coherent and tied the current moves to max moves better. Yeah, some of the string parameters are a bit interesting but I decided I didn't want to start messing with a bunch of types as they are passed through down to this level.

RoiEXLab commented on game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java:
```diff
@@ -568,7 +570,9 @@ public void paint(final Graphics g) {
     }
     if (routeDescription != null) {
-      routeDrawer.drawRoute(g2d, routeDescription, movementLeftForCurrentUnits);
+      routeDrawer.drawRoute(g2d, routeDescription, movementLeftForCurrentUnits, movementFuelCost,
```
It would be nice if drawRoute would continue to only accept 3 arguments, the third one being a wrapper object to wrap the last 3 objects.
This would keep the related stuff together and the parameter declarations relatively clean.

ron-murhammer replied:
I think as a whole there could be some refactoring of parameters across a lot of the MapRouteDrawer and MapPanel. Both in terms of wrapper objects and types used. I decided not to dive into that for this PR. I think looking at even existing objects like RouteDescription which might be expanded to include these. But for now, I think that it would be better in a separate PR.

ron-murhammer commented on game-core/src/test/java/games/strategy/triplea/ui/RouteTest.java:
```diff
@@ -83,7 +87,10 @@ public void testCorrectParameterHandling() {
     when(mockShape.contains(any(Point2D.class))).thenReturn(true);
-    routeDrawer.drawRoute(mockGraphics, dummyRouteDescription, "2");
+    final ResourceCollection mockResourceCollection = mock(ResourceCollection.class);
+    when(mockResourceCollection.getResourcesCopy()).thenReturn(new IntegerMap<Resource>());
```
Yep, no idea why I put the type. Updated.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java:
```diff
@@ -89,6 +90,7 @@
   private final TileManager tileManager;
   private BufferedImage mouseShadowImage = null;
   private String movementLeftForCurrentUnits = "";
+  private ResourceCollection movementFuelCost;
```
Good point. It doesn't seem to ever happen but not sure I can prove it isn't possible. Updated.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -314,28 +323,81 @@ private void drawCurvedPath(final Graphics2D graphics, final Point2D[] points) {
+    final String unitMovementLeft = (maxMovement == null || maxMovement.trim().length() == 0) ? "" : "/" + maxMovement;
+    final int imageWidth = findMovementLeftImageWidth(curMovement, unitMovementLeft, movementFuelCost);
+    final BufferedImage image = new BufferedImage(imageWidth, MESSAGE_HEIGHT, BufferedImage.TYPE_INT_ARGB);
+    final Graphics2D graphics = image.createGraphics();
```
Good point. Seems that some of the existing graphics code does use dispose() and other places it doesn't. Something that could probably be improved. Other areas seem to just be calling dispose() at the end of the usage without a try-finally. I think that is probably ok given this is as important as say closing a stream. I'll stick with that pattern for now for consistency and we can always revisit adding dispose() to all graphics areas and whether try-finally would be better.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/ui/MapRouteDrawer.java:
```diff
@@ -314,28 +323,81 @@ private void drawCurvedPath(final Graphics2D graphics, final Point2D[] points) {
+
+    // Create temp graphics to calculate necessary image width based on font sizes
+    final BufferedImage tempImage = new BufferedImage(1, 1, BufferedImage.TYPE_INT_ARGB);
+    final Graphics2D tempGraphics = tempImage.createGraphics();
```
Updated.

ron-murhammer commented:
@ssoloff @RoiEXLab Should have addressed all the comments. I'm open to thoughts on colors though don't want to end up in a rabbit hole of trying tons of different things as we can always adjust based on community feedback.

ron-murhammer commented:
Gonna go ahead and merge this so testing can begin.
